src/modules/assessments/components/AssessmentForm.tsx

```diff
@@ -124,6 +125,13 @@ const AssessmentForm = ({
       },
     });
 
+  const handleDirty = useCallback(
+    (value: boolean) => {
+      onDirty?.(enrollment.id, value);
+    },
+    [onDirty, enrollment.id]
+  );
+
```

Comment:
* I modified AssessmentForm so that the useAssessmentHandlers callbacks unset the saving state. I think that fixes the first issue. I should add a test case for this partially submitted assessment workflow
* That's a great find with Re the save/next sequence. I was able to reproduce that reliably and got a fix in place. The way I addressed it was to add a 'nextTab' state that pauses navigation until the save completes.
* The above requires tracking both in-flight requests and dirty state on the form. I tried to also add some state to track errors but I couldn't get it behave reliably and was concerned it was taking too long, particularly if we are going to refactor it.

Reply:
Awesome, I think this looks good. Thank you. I share the concern re: time - I think it's worth prioritizing a refactor here, since there are additional features/enhancements that are going to be very difficult (time consuming) to tack on to the current approach.